iothub/device/src/InternalClient.cs

```diff
@@ -1835,10 +1753,11 @@ public async Task SetMessageHandlerAsync(MessageHandler messageHandler, object u
                 Logging.Enter(this, messageHandler, userContext, nameof(SetMessageHandlerAsync));
             }
 
+            cancellationToken.ThrowIfCancellationRequested();
+            await _moduleReceiveMessageSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
             try
             {
-                cancellationToken.ThrowIfCancellationRequested();
-                await _moduleReceiveMessageSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
As per the model I described in the description of this PR, grabbing the semaphore and checking the cancellation token must be outside of the try/finally block to ensure that we don't try to release a semaphore that we never grabbed.